kawin/Diffusion.py

```diff
@@ -0,0 +1,825 @@
+        if toCSV:
+            headers = ['Distance(m)']
+            arrays = [self.z]
+            for i in range(len(self.allElements)):
+                headers.append('x(' + self.allElements[i] + ')')
+                if i == 0:
+                    arrays.append(1 - np.sum(self.x, axis=0))
+                else:
+                    arrays.append(self.x[i-1,:])
+            for i in range(len(self.phases)):
+                headers.append('f(' + self.phases[i] + ')')
+                arrays.append(self.p[i,:])
+            rows = zip_longest(*arrays, fillvalue='')
+            if '.csv' not in filename.lower():
+                filename = filename + '.csv'
+            with open(filename, 'w', newline='') as f:
+                csv.writer(f).writerow(headers)
+                csv.writer(f).writerows(rows)
+        else:
+            variables = ['zlim', 'N', 'allElements', 'phases', 'z', 'x', 'p']
+            vDict = {v: getattr(self, v) for v in variables}
+            if compressed:
+                np.savez_compressed(filename, **vDict, allow_pickle=True)
+            else:
+                np.savez(filename, **vDict, allow_pickle=True)
+
+    def load(filename):
+        if '.np' in filename.lower():
+            data = np.load(filename, allow_pickle=True)
+            model = DiffusionModel(data['zlim'], data['N'], data['allElements'], data['phases'])
+            model.z = data['z']
+            model.x = data['x']
+            model.p = data['p']
+        else:
+            with open(filename, 'r') as csvFile:
+                data = csv.reader(csvFile, delimiter=',')
+                i = 0
+                headers = []
+                columns = {}
+                for row in data:
+                    if i == 0:
+                        headers = row
+                        columns = {h: [] for h in headers}
+                    else:
+                        for j in range(len(row)):
+                            if row[j] != '':
+                                columns[headers[j]].append(float(row[j]))
+                    i += 1
+            
+            elements, phases = [], []
+            x, p = [], []
+            for h in headers:
+                if 'Distance' in h:
+                    z = columns[h]
+                elif 'x' in h:
+                    elements.append(h[2:-1])
+                    x.append(columns[h])
+                elif 'f' in h:
+                    phases.append(h[2:-1])
+                    p.append(columns[h])
+            model = DiffusionModel([z[0], z[-1]], len(z), elements, phases)
+            model.z = np.array(z)
+            model.x = np.array(x)[1:,:]
+            model.p = np.array(p)
+        return model   
+
+    def setHashSensitivity(self, s):
+        self.hashSensitivity = np.power(10, int(s))
+
+    def _getHash(self, x):
+        return hash(tuple((x*self.hashSensitivity).astype(np.int32)))
+        #return int(np.sum(np.power(self.hashSensitivity, 1+np.arange(len(x))) * x))
```

Comment:
I realized I forgot to put docstrings on some of the functions in the diffusion model, including these two functions. These will be added for all the functions in Diffusion.py.

The hash table is somewhat of a caching method that is briefly mentioned in the homogenization model paper to reduce the number of equilibrium calculations and thus the simulation time. The idea is to store computed equilibrium values in a hash table with the hashes based off the composition. Then if a composition of a volume element is very close to a composition that was seen previously, then use the already computed equilibrium values rather than computing equilibrium again.

The sensitivity determines how many significant figures to store when creating a hash from a composition. For example, using a sensitivity of 3, a composition of (a=0.5234, b=0.4776) and (a=0.5232, b=0.4778) will both give the same hash value as (a=0.523, b=0.478) would. A lower sensitivity would mean that a larger range of compositions will create the same hash and thus, less equilibrium calculations will need to be done and the simulation will be faster. But this will also reduce the accuracy of the simulation. The opposite will be true for a high sensitivity. By default, the sensitivity is set to 4, which I found to work well for the systems I tried the diffusion models on so far.

In practice, fluxes will be calculated like this for each volume element:
1. Compute hash for composition
2. Check if hash already exists in hash table, if not, compute equilibrium and add to hash table
3. Grab equilibrium value from hash table
4. Compute flux from equilibrium value

Although the hash table was mentioned in the homogenization paper, this is also implemented for the single phase model, where instead of storing equilibrium values, the diffusivity will be stored.